### PR TITLE
Handle module cycles

### PR DESCRIPTION
**Pull request by grouma**
- Introduce new MetaModuleCleanBuilder which outputs a .meta_module_clean
- This new output contains a list of cleaned modules with cyclic modules merged into one

I'm seeing even further performance benefits with this approach. Curious how it will impact the Wrike folks.

Towards https://github.com/dart-lang/build/issues/1186

### CODE REVIEW

File: build_modules/build.yaml [hunk 1]
```diff
@@ -1,21 +1,15 @@
...
     import: "package:build_modules/builders.dart"
     builder_factories:
+      - metaModuleBuilder
+      - metaModuleCleanBuilder
       - moduleBuilder
       - unlinkedSummaryBuilder
       - linkedSummaryBuilder
     build_extensions:
+      $lib$:
+        - .meta_module
```
Comment (natebosch) [inline build_modules/build.yaml]:
```yaml
- .meta_module.raw
- .meta_module.clean
```

?

File: build_modules/lib/src/meta_module_clean_builder.dart [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import 'dart:convert';
+
+import 'package:graphs/graphs.dart';
+
+import 'package:build/build.dart';
+import 'meta_module.dart';
+import 'meta_module_builder.dart';
+import 'modules.dart';
+
+/// The extension for serialized clean meta module assets.
```
Comment (natebosch) [inline build_modules/lib/src/meta_module_clean_builder.dart]: I'm unsure about the name "clean"...

Maybe `.package_modules` and `.cross_package_modules` instead?

File: build_modules/lib/src/meta_module_clean_builder.dart [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  nextModules.addAll(meta.modules);
+  while (nextModules.isNotEmpty) {
+    var module = nextModules.removeLast();
+    dependentModules.add(module);
+    for (var source in module.sources) {
+      assetToModule[source] = module;
+    }
+    for (var dep in module.directDependencies) {
+      var depMetaAsset = new AssetId(dep.package, 'lib/$metaModuleExtension');
+      if (!seenMetas.contains(depMetaAsset)) {
```
Comment (natebosch) [inline build_modules/lib/src/meta_module_clean_builder.dart]: reduce nesting by one by reversing this condition with a `continue`

File: build_modules/lib/src/meta_module_clean_builder.dart [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+const metaModuleCleanExtension = '.meta_module_clean';
+
+/// Map of [AssetId] to corresponding primary [AssetId] within the same
+/// [Module].
+final _assetToModule =
+    new Resource<Map<AssetId, Module>>(() => {}, dispose: (map) => map.clear());
+
+/// Returns a set of all modules transitvely reachable from the provided meta
+/// module asset.
+Future<Set<Module>> _transitiveModules(BuildStep buildStep, AssetId metaAsset,
```
Comment (natebosch) [inline build_modules/lib/src/meta_module_clean_builder.dart]: I wonder how well this fits with `crawlAsync`?

https://pub.dartlang.org/documentation/graphs/0.1.1/graphs/crawlAsync.html

The benefit would be that we'd separate out the business logic of how we resolve modules and dependencies from the graph algorithm

File: build_modules/lib/src/meta_module_clean_builder.dart [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  Future build(BuildStep buildStep) async {
+    var assetToModule = await buildStep.fetchResource(_assetToModule);
+    var modules =
+        await _transitiveModules(buildStep, buildStep.inputId, assetToModule);
+    var connectedComponents = stronglyConnectedComponents<AssetId, Module>(
+        modules,
+        (m) => m.primarySource,
+        (m) => m.directDependencies
+            .map((d) => assetToModule[d])
+            .where((d) => d != null));
```
Comment (natebosch) [inline build_modules/lib/src/meta_module_clean_builder.dart]: when would we not have a module in `assetToModule`?

File: build_modules/lib/src/meta_module_clean_builder.dart [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    // Ensure deterministic output by sorting the modules.
+    var cleanModules = new SplayTreeSet<Module>(
+        (a, b) => a.primarySource.compareTo(b.primarySource));
+    for (var connectedComponent in connectedComponents) {
```
Comment (natebosch) [inline build_modules/lib/src/meta_module_clean_builder.dart]: could write this in a way that more directly expresses that we're interested the collection rather than side effects.

```dart
  Module merge(List<Module> c) => _mergeComponent(c, assetToModule);
  bool hasSourceInPackage(Module m) =>
      m.sources.any((s) => s.package == buildStep.inputId.package);
  new SplayTreeSet<Module>()
    ..addAll(connectedComponents.map(merge).where(hasSourceInPackage));
```

File: build_modules/lib/src/meta_module_clean_builder.dart [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+/// The extension for serialized clean meta module assets.
+const metaModuleCleanExtension = '.meta_module_clean';
+
+/// Map of [AssetId] to corresponding primary [AssetId] within the same
+/// [Module].
+final _assetToModule =
+    new Resource<Map<AssetId, Module>>(() => {}, dispose: (map) => map.clear());
+
+/// Returns a set of all modules transitvely reachable from the provided meta
```
Comment (jakemac53) [inline build_modules/lib/src/meta_module_clean_builder.dart]: transitively*

File: build_modules/lib/src/meta_module_clean_builder.dart (hunk 4 above)
Comment (jakemac53) [reply build_modules/lib/src/meta_module_clean_builder.dart]: You could do a `crawlAsync` for `MetaModule`s (children would be all the merged MetaModule dependencies of all the contained modules), and then you could expand all the modules afterwords?

Comment (jakemac53) [inline build_modules/lib/src/meta_module_clean_builder.dart]: nit: I prefer to have private implementations below the public apis in the file. Also I think its a bit more natural to put private methods implementations below their actual usage in the file, roughly in the order in which they are invoked by the public methods.

File: build_modules/lib/src/meta_module_clean_builder.dart [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  var deps = new Set<AssetId>();
+  // Sort the modules to deterministicly select the primary source.
+  var components = new SplayTreeSet<Module>(
+      (a, b) => a.primarySource.compareTo(b.primarySource))
+    ..addAll(connectedComponent);
+  var primarySource = components.first.primarySource;
+  for (var module in connectedComponent) {
+    sources.addAll(module.sources);
+    for (var dep in module.directDependencies) {
+      var depModule = assetToModule[dep];
```
Comment (jakemac53) [inline build_modules/lib/src/meta_module_clean_builder.dart]: It is a bit hard to follow, but will this map contain un-merged modules? We need to make sure it ends up pointing at the merged modules returned by this function, right?

**Comment by jakemac53** [comment]
I would really like to see some full E2E tests to confirm all the cases we care about are covered, and all the interactions work as we expect. At least these two cases:

- A root package cycle should make all packages in the cycle revert to the fine strategy.
- An example of a cycle in package dependencies.

File: build_modules/lib/src/meta_module_clean_builder.dart (hunk 4 above)
Comment (grouma) [reply build_modules/lib/src/meta_module_clean_builder.dart]: I can't use the crawlAsync method since it doesn't propagate errors as expected.

File: build_modules/lib/src/meta_module_clean_builder.dart (hunk 2 above)
Comment (natebosch) [reply build_modules/lib/src/meta_module_clean_builder.dart]: Can the doc comment explain what "clean" means?

File: build_modules/lib/src/meta_module_clean_builder.dart (hunk 4 above)
Comment (natebosch) [reply build_modules/lib/src/meta_module_clean_builder.dart]: https://github.com/dart-lang/graphs/pull/18

File: build_modules/lib/src/meta_module_clean_builder.dart (hunk 2 above)
Comment (grouma) [reply build_modules/lib/src/meta_module_clean_builder.dart]: Done.
